execution.py

```diff
@@ -73,6 +105,26 @@ def slice_dict(d, i):
             if allow_interrupt:
                 nodes.before_node_execution()
             results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
+
+    #Ensure IMAGE outputs conform to BHWC
+    return_indexs = {}
+    formated_results = []
+    
+    if hasattr(obj, "RETURN_NAMES") and hasattr(obj, "RETURN_TYPES"):       
+        for i, t in enumerate(obj.RETURN_TYPES):
+            return_indexs[i] = t
+        
+        for i, r in enumerate(results[0]):
+            if return_indexs[i] == "IMAGE":
```

Comment:
Instead of special-casing this for `IMAGE`s, could we have a dictionary of type validation functions? I don't imagine that any other validation functions would actually modify the output, but it would be nice to warn people that their `INT` is actually a `STRING` (or visa-versa) so we don't run into this issue (where some nodes work and some don't due to an invalid type) again in the future.

Reply:
I agree, we can ensure type conformity for all specified types. I will have to look again, but in some places if you type it as FLOAT for example, in the files with typing its typed as int|float|string, so I think its probably possible to pass them to the wrong types if you try really hard. I haven't personally run into this experience, but the validation function can be abstracted. Do you know if that actually happens?

Reply:
Is `FLOAT` really typed as `int|float|string` somewhere? I agree that INT and FLOAT are probably interchangeable and we don't need to warn when one is used instead of the other, but plenty of nodes are going to break if you return a string instead of a number.

Reply:
<img width="303" alt="Screenshot 2024-07-30 at 11 23 34 PM" src="https://github.com/user-attachments/assets/521fc9cc-b0c8-4e39-9962-54674eb3a8ec">
In the app.js yes. 

In the actual node, 3: Float, 4: Float, 5: Int, 6: Int

Reply:
Oh, that's a node using a variant type though. It's supported by litegraph, but not officially supported by ComfyUI. I would ignore those types for the purpose of validation. (We don't have to explicitly ignore them -- since `"FLOAT,INT,NUMBER,STRING"` looks like one type to ComfyUI, it won't match the dictionary.)

I actually wouldn't do any conversion. Long term, I think we should be looking to remove even the conversion added in this PR. Instead, I would focus on spitting out warnings to the console (so node authors fix their nodes). After some period of time, we can change those into errors so that node authors see it the very first time they try their node.

So the validation dictionary would be something like:
```python
def generate_type_validator(valid_types):
    def validate_type(node_type output_name, value):
        if not isinstance(value, valid_types):
            # In 6 months, replace this print statement with raising an error
            print(f"The {output_name} output of the {node_type} is an invalid type. [Probably more info here]")
        # Don't actually change anything
        return value
    return validate_type

validation_funcs = {
    "IMAGE": force_bhw3, # Your existing function (updated to take the input name for logging purposes)
    "INT": generate_type_validator((int, float)),
    "FLOAT": generate_type_validator((float, int)),
    "STRING": generate_type_validator((str,)),
    # etc.
}
```

Reply:
I quite like that validation generator.

However, I do think type conversion should maybe be done, given the way it appears from what Ive looked at today. I honestly would not be surprised if you can pass a string or an int, or a float, through the IMAGE pipe, or an Image through the string pipe, or any other pipe. The only thing that seems to be getting done, is specifying the type for the UI and only letting same "typed" pipes connect, it doesn't seem to have any actual bearing on what type of data can be passed through however, much less how it should be formatted.

That all the type information is discarded for the outputs, and only "re-hinted" with no enforcement on input tells me that something typed as `INT` should only be able to carry and `INT`, `FLOAT`, a `FLOAT`

I did successfully pass a `STRING` through the image pipe so I'll have to take a deeper look at all of this tomorrow. It looks like BHW3 not being enforced on `IMAGE` was just a sign that no typing is enforced on anything. Non-typed non-sanitized inputs is a security problem waiting to happen.

I just successfully passed a `function` and a `class` through the `IMAGE` pipe. All it needs a .shape method that can execute arbitrary code when called or any tensor operations and code can hitch a ride.

knowing this, I now strongly believe this needs to be type enforced, not just warnings. 

<img width="442" alt="Screenshot 2024-07-31 at 12 57 36 AM" src="https://github.com/user-attachments/assets/4eaef115-4888-4058-85fe-78ad9e1ccf58">

<img width="679" alt="Screenshot 2024-07-31 at 12 52 25 AM" src="https://github.com/user-attachments/assets/52b2b83b-dfff-41e7-a7f6-29bac3f878a8">

<img width="475" alt="Screenshot 2024-07-31 at 12 57 08 AM" src="https://github.com/user-attachments/assets/bf976cff-5909-4be7-952c-a1ff45afd27d">

its completely possible to pass arbitrary functions/classes along any pipe with payloads and not cause any errors as long as you mirror any inbuilt operations that will be called. 

I now thinki that IMAGE needs to be forced to conform to torch.tensor, and other types need to conform to their specified types, not just warnings. 

The channel layout for IMAGE yeah, let that be a warning item, but IMAGE passing something that is not torch.tensor, should not work.